### PR TITLE
TINY-8235: Convert the silver theme to use the new options API

### PR DESCRIPTION
**Pull request by lnewson**
Related Ticket: TINY-8235

Description of Changes:

List PR in the set of 6 PRs to convert all the plugins/themes to use the new Options API. This also updates McAgar to work with the new Options API as without this tests fail.

Note: I also realised we need to keep the mobile specific overrides in `NormalizeOptions` because we want them to always override the desktop options. So if someone sets say `toolbar_mode: 'sliding'` in the main options it should remain `scrolling` on mobile unless overridden in the `mobile` config section.

Pre-checks:
* [x] ~Changelog entry added~
* [x] ~Tests have been added (if applicable)~
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/mcagar/src/main/ts/ephox/mcagar/alien/Settings.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { Editor } from './EditorTypes';
+
+const getSetting = <R>(editor: Editor, name: string): R => {
```
Comment (spocke) [inline modules/mcagar/src/main/ts/ephox/mcagar/alien/Settings.ts:3]: Maybe it will be just confusing to any future developer that these functions are called `getSetting` since the options name is the way forward.

File: modules/mcagar/src/main/ts/ephox/mcagar/api/pipeline/ApiChains.ts [hunk 2]
```diff
@@ -57,13 +58,13 @@ const cSetSelection = <T extends Editor>(startPath: number[], soffset: number, f
 
 const cSetSetting = <T extends Editor>(key: string, value: any): Chain<T, T> => {
```
Comment (spocke) [inline modules/mcagar/src/main/ts/ephox/mcagar/api/pipeline/ApiChains.ts:59]: Should we deprecate these and add new cSetOption and setOption to mcagar and over time switch to that?

File: modules/tinymce/src/core/main/ts/api/Options.ts [hunk 3]
```diff
@@ -646,6 +655,11 @@ const register = (editor: Editor) => {
     default: false
   });
 
+  registerOption('content_editable_state', {
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/api/Options.ts:658]: Maybe add a todo about this going away or just remove it now?

File: modules/tinymce/src/plugins/autoresize/main/ts/Plugin.ts [hunk 4]
```diff
@@ -23,10 +23,6 @@ import * as Resize from './core/Resize';
 export default (): void => {
   PluginManager.add('autoresize', (editor) => {
     Options.register(editor);
-    // If autoresize is enabled, disable resize if the user hasn't explicitly enabled it
```
Comment (spocke) [inline modules/tinymce/src/plugins/autoresize/main/ts/Plugin.ts:26]: Curious why was this removed? Will this render a resize handle when the autoresize plugin is enabled or is all that now part of the theme and this is no longer needed?

File: modules/tinymce/src/themes/silver/main/ts/ui/core/color/Options.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
+/**
```
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/core/color/Options.ts:1]: I think if you rename files do that before you do other changes in the same commit and it should be easier to diff. If more than 70% of the file is changed it not picked up as a rename. But I think if you do it in two steps it is.

File: modules/tinymce/src/themes/silver/main/ts/ui/statusbar/ResizeHandle.ts [hunk 6]
```diff
@@ -11,14 +11,14 @@ import { SugarPosition } from '@ephox/sugar';
...
 import Editor from 'tinymce/core/api/Editor';
 
+import * as Options from '../../api/Options';
 import { UiFactoryBackstageProviders } from '../../backstage/Backstage';
 import * as Icons from '../icons/Icons';
 import { resize, ResizeTypes } from '../sizing/Resize';
 
 const getResizeType = (editor: Editor): ResizeTypes => {
   // If autoresize is enabled, disable resize
-  const fallback = !editor.hasPlugin('autoresize');
```
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/statusbar/ResizeHandle.ts:20]: What effects does this have on resizing and the autoresize plugin?

**Review by spocke — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/api/Options.ts (hunk 3 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/Options.ts:658]: We have a separate Jira to clean this up so don't want to touch it just yet.

File: modules/tinymce/src/plugins/autoresize/main/ts/Plugin.ts (hunk 4 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/autoresize/main/ts/Plugin.ts:26]: Logic already existed in Silver for this so I just removed this as it doesn't make sense to have it in 2 places.

File: modules/tinymce/src/themes/silver/main/ts/ui/statusbar/ResizeHandle.ts (hunk 6 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/statusbar/ResizeHandle.ts:20]: Oops the above comment was meant to be removed as well... it has no difference on functionality as this just moved to the default in when we register the option.

File: modules/tinymce/src/themes/silver/main/ts/ui/statusbar/ResizeHandle.ts [hunk 7]
```diff
@@ -11,14 +11,14 @@ import { SugarPosition } from '@ephox/sugar';
 
 import Editor from 'tinymce/core/api/Editor';
 
+import * as Options from '../../api/Options';
 import { UiFactoryBackstageProviders } from '../../backstage/Backstage';
 import * as Icons from '../icons/Icons';
 import { resize, ResizeTypes } from '../sizing/Resize';
 
 const getResizeType = (editor: Editor): ResizeTypes => {
   // If autoresize is enabled, disable resize
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/statusbar/ResizeHandle.ts:20]:
```suggestion
```

**Review by metricjs — APPROVED** [review]

File: modules/mcagar/src/main/ts/ephox/mcagar/api/pipeline/ApiChains.ts (hunk 2 above)
Comment (lnewson) [reply modules/mcagar/src/main/ts/ephox/mcagar/api/pipeline/ApiChains.ts:59]: It's already listed in the `cleanup` settings Jira that's meant to be done after everything is converted. We probably should just mention to rename it there as I really didn't want to touch McAgar here but had to unfortunately... At the moment it's still very much geared more at 4.x/5.x than at 6.x.

File: modules/tinymce/src/core/main/ts/api/Options.ts [hunk 8]
```diff
@@ -369,6 +369,15 @@ const register = (editor: Editor) => {
...
   });
 
+  registerOption('skin', {
+    processor: (value) => Type.isString(value) || value === false,
+    default: 'oxide'
+  });
+
+  registerOption('skin_url', {
+    processor: 'string'
+  });
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/Options.ts:379]: Forgot to mention this when I logged the PR but I was tossing up between whether this should be in Core or in the Silver Theme. Would be keen to get others thoughts?

Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/api/Options.ts:379]: Ooh, tough choice. It depends on whether the skin setting is ever read in core (other than to download the skin, although perhaps that counts).

The skin is not useful without the theme, so I suggest putting the registration in the theme, unless that would cause problems with downloading the skin.

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/Options.ts:379]: The skin option I don't believe is ever read in core, as it's loaded by the theme. The reason I placed it in core is it's used by other plugins (e.g. template, preview and exports)

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/Options.ts:379]: Anyways yeah I really just flipped a coin here... I do agree thinking more on this that it should be in the theme, as each theme may have a different default skin.

**Review by TheSpyder — APPROVED** [review]

File: modules/mcagar/src/main/ts/ephox/mcagar/alien/Settings.ts (hunk 1 above)
Comment (lnewson) [reply modules/mcagar/src/main/ts/ephox/mcagar/alien/Settings.ts]: Have renamed this and deprecated the old functions

File: modules/mcagar/src/main/ts/ephox/mcagar/api/pipeline/ApiChains.ts (hunk 2 above)
Comment (lnewson) [reply modules/mcagar/src/main/ts/ephox/mcagar/api/pipeline/ApiChains.ts]: I did this anyways since it was quick...

File: modules/tinymce/src/plugins/autoresize/main/ts/Plugin.ts (hunk 4 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/autoresize/main/ts/Plugin.ts:26]: I've reinstated this but as we found out today it's never done anything since 5.0. The reason being the theme is looking up this and rendering the resize handle before the plugins initialise. I've logged TINY-8288 to fix that at some stage.

File: modules/tinymce/src/themes/silver/main/ts/ui/core/color/Options.ts (hunk 5 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/core/color/Options.ts:1]: Good to know, I wasn't aware of that so I'll try and remember to give it a go in future.
